File: frontend/types/media.ts
/**
 * TrueSight — TypeScript Types: Media
 * Represents uploaded media files and their metadata.
 */

/** Supported media types for analysis. */
export type MediaType = "IMAGE" | "AUDIO" | "VIDEO" | "URL";

/** Allowed file extensions per media type. */
export const ALLOWED_IMAGE_EXTENSIONS = ["jpg", "jpeg", "png", "webp", "gif"] as const;
export const ALLOWED_AUDIO_EXTENSIONS = ["mp3", "wav", "m4a", "ogg"] as const;
export const ALLOWED_VIDEO_EXTENSIONS = ["mp4", "mov", "webm", "avi"] as const;

export type ImageExtension = (typeof ALLOWED_IMAGE_EXTENSIONS)[number];
export type AudioExtension = (typeof ALLOWED_AUDIO_EXTENSIONS)[number];
export type VideoExtension = (typeof ALLOWED_VIDEO_EXTENSIONS)[number];

/** Maximum allowed upload size: 50 MB. */
export const MAX_UPLOAD_SIZE_BYTES = 52_428_800;

/** A media item stored in Supabase Storage. */
export interface MediaItem {
  id: string;
  userId: string;
  analysisId: string | null;
  filename: string;
  originalFilename: string;
  contentType: string;
  sizeBytes: number;
  mediaType: MediaType;
  storagePath: string;
  /** Signed URL generated per-request; not persisted. */
  signedUrl?: string;
  createdAt: string;
}

/** Helper to determine media type from a File object. */
export function inferMediaType(file: File): MediaType | null {
  const ext = file.name.split(".").pop()?.toLowerCase();
  if (!ext) return null;
  if ((ALLOWED_IMAGE_EXTENSIONS as readonly string[]).includes(ext)) return "IMAGE";
  if ((ALLOWED_AUDIO_EXTENSIONS as readonly string[]).includes(ext)) return "AUDIO";
  if ((ALLOWED_VIDEO_EXTENSIONS as readonly string[]).includes(ext)) return "VIDEO";
  return null;
}

File: backend/app/schemas/media.py
"""
TrueSight Backend — Pydantic Schemas: Media
Models representing uploaded media files.
"""
from __future__ import annotations

from datetime import datetime
from typing import Optional
from uuid import UUID

from pydantic import BaseModel, Field

from app.schemas.analysis import MediaType


class MediaItem(BaseModel):
    """Represents a single uploaded media item stored in Supabase Storage."""

    id: UUID
    user_id: UUID
    analysis_id: Optional[UUID] = None

    filename: str
    original_filename: str
    content_type: str
    size_bytes: int
    media_type: MediaType

    # Supabase Storage path: media/{user_id}/{analysis_id}/{filename}
    storage_path: str

    # Signed URL is generated per-request; not stored permanently.
    signed_url: Optional[str] = None

    created_at: datetime


# ── Allowed file types ────────────────────────────────────────────────────────

ALLOWED_IMAGE_TYPES: frozenset[str] = frozenset(
    {"image/jpeg", "image/png", "image/webp", "image/gif"}
)

ALLOWED_AUDIO_TYPES: frozenset[str] = frozenset(
    {"audio/mpeg", "audio/wav", "audio/x-wav", "audio/mp4", "audio/ogg"}
)

ALLOWED_VIDEO_TYPES: frozenset[str] = frozenset(
    {"video/mp4", "video/quicktime", "video/webm", "video/x-msvideo"}
)

ALLOWED_EXTENSIONS: frozenset[str] = frozenset(
    {"jpg", "jpeg", "png", "webp", "gif", "mp3", "wav", "m4a", "ogg", "mp4", "mov", "webm", "avi"}
)

ALL_ALLOWED_CONTENT_TYPES: frozenset[str] = (
    ALLOWED_IMAGE_TYPES | ALLOWED_AUDIO_TYPES | ALLOWED_VIDEO_TYPES
)
